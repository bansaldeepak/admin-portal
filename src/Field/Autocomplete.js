import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

import TextField from '@material-ui/core/TextField';
import Autocomplete from '@material-ui/lab/Autocomplete';
import CircularProgress from '@material-ui/core/CircularProgress';

const useStyles = makeStyles((theme) => ({
  textField: {
    minWidth: '100px'
  }
}));

export default function EnhancedAutocomplete (props) {
  const classes = useStyles();
  const { minlength, key, name, optionKey, optionValue, handleChange } = props;
  const [open, setOpen] = React.useState(false);
  const [options, setOptions] = React.useState([]);
  const [value, setValue] = React.useState();
  // const loading = open && options.length === 0;
  const loading = open;

  React.useEffect(() => {
    let active = true;

    if (!loading) {
      return undefined;
    }

    (async () => {
      let resultOptions = [];
      if (typeof props.options === 'object') {
        if (value) {
          resultOptions = props.options.filter((obj) => {
            return Object.keys(obj).reduce((acc, curr) => {
              return acc || obj[curr].toLowerCase().includes(value);
            }, false);
          });
        } else {
          resultOptions = props.options;
        }
      } else if (typeof props.options === 'function') {
        if (value && value.length >= (minlength || 3)) {
          resultOptions = await props.options(value);
        }
      }

      if (active) {
        setOptions(resultOptions);
      }
    })();

    return () => {
      active = false;
    };
  }, [loading, value]);

  return (
    <Autocomplete
      key={key}
      name={name}
      open={open}
      onOpen={() => {
        setOpen(true);
      }}
      onClose={() => {
        setOpen(false);
      }}
      onChange={(event, value) => {
        const selectedValue = value[optionKey] || value[name] || value.id || value.key || value.value || value.name || value;
        handleChange(selectedValue);
      }}
      getOptionSelected={(option, selected) => (option[optionKey] || option[name] || option.id || option.key || option.value || option.name || option) === (selected[optionKey] || selected[name] || selected.id || selected.key || selected.value || selected.name || selected)}
      getOptionLabel={(option) => option[optionValue] || option.name || option.label || option.value || option}
      options={options}
      loading={loading}
      renderInput={(params) => {
        return (
          <TextField
            {...params}
            className={classes.textField}
            label={props.label}
            onChange={(event) => {
              const { value } = event.target;
              setValue(value);
            }}
            InputProps={{
              ...params.InputProps,
              endAdornment: (
                <React.Fragment>
                  {loading ? <CircularProgress color='inherit' size={20} /> : null}
                  {params.InputProps.endAdornment}
                </React.Fragment>
              )
            }}
          />
        );
      }}
      filterOptions={(options) => {
        return options;
      }}
    />
  );
};

EnhancedAutocomplete.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({}))
};

EnhancedAutocomplete.defaultProps = {
  options: null
};
