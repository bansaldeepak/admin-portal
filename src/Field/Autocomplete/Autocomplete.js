import React from "react";
import PropTypes from "prop-types";
import { makeStyles } from "@material-ui/core/styles";

import TextField from "@material-ui/core/TextField";
import Autocomplete from "@material-ui/lab/Autocomplete";
import CircularProgress from "@material-ui/core/CircularProgress";

const useStyles = makeStyles((theme) => ({
  textField: {
    minWidth: "100px",
  },
}));

import { isEmpty } from "../../lib";

const EnhancedAutocomplete = (props) => {
  const classes = useStyles();
  const {
    minlength,
    key,
    name,
    optionKey,
    optionValue,
    defaultValue,
    handleChange,
  } = props;

  const [open, setOpen] = React.useState(false);
  const [options, setOptions] = React.useState([]);
  const [value, setValue] = React.useState();
  // const loading = open && options.length === 0;
  const loading = open;
  React.useEffect(() => {
    let active = true;

    if (!loading) {
      return undefined;
    }

    (async () => {
      let resultOptions = [];
      if (typeof props.options === "object") {
        if (!isEmpty(value)) {
          resultOptions = props.options.filter((obj) => {
            return obj.toLowerCase().includes(value.toLowerCase());
          });
        } else {
          resultOptions = props.options;
        }
      } else if (typeof props.options === "function") {
        if (value && value.length >= minlength) {
          resultOptions = await props.options(value.toLowerCase());
        }
      }

      if (active) {
        setOptions(resultOptions);
      }
    })();

    return () => {
      active = false;
    };
  }, [loading, value]);

  return (
    <Autocomplete
      key={key}
      name={name}
      open={open}
      defaultValue={defaultValue}
      onOpen={() => {
        setOpen(true);
      }}
      onClose={() => {
        setOpen(false);
      }}
      onChange={(event, value) => {
        const selectedValue = !isEmpty(value)
          ? value[optionKey] ||
            value[name] ||
            value.id ||
            value.key ||
            value.value ||
            value.name ||
            value
          : "";
        handleChange(selectedValue);
      }}
      getOptionSelected={(option, selected) =>
        (option[optionKey] ||
          option[name] ||
          option.id ||
          option.key ||
          option.value ||
          option.name ||
          option) ===
        (selected[optionKey] ||
          selected[name] ||
          selected.id ||
          selected.key ||
          selected.value ||
          selected.name ||
          selected)
      }
      getOptionLabel={(option) =>
        option[optionValue] ||
        option.name ||
        option.label ||
        option.value ||
        option
      }
      options={options}
      loading={loading}
      renderInput={(params) => {
        return (
          <TextField
            {...params}
            className={classes.textField}
            label={props.label}
            onChange={(event) => {
              const { value } = event.target;
              setValue(value);
            }}
            InputProps={{
              ...params.InputProps,
              endAdornment: (
                <React.Fragment>
                  {loading ? (
                    <CircularProgress color="inherit" size={20} />
                  ) : null}
                  {params.InputProps.endAdornment}
                </React.Fragment>
              ),
            }}
          />
        );
      }}
      filterOptions={(options) => {
        return options;
      }}
    />
  );
};

EnhancedAutocomplete.propTypes = {
  options: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.arrayOf(PropTypes.string),
    PropTypes.arrayOf(PropTypes.shape({})),
  ]),
  minlength: PropTypes.number,
  key: PropTypes.string,
  name: PropTypes.string,
  optionKey: PropTypes.string,
  optionValue: PropTypes.string,
  handleChange: PropTypes.func,
  /**
   * Default value for the field
   */
  defaultValue: PropTypes.string,
};

EnhancedAutocomplete.defaultProps = {
  options: [],
  minlength: 3,
  optionKey: "key",
  optionValue: "value",
  handleChange: () => {},
  defaultValue: "",
};

export default EnhancedAutocomplete;
