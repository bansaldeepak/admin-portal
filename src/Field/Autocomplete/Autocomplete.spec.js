import React from "react";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { shallowToJson } from "enzyme-to-json";
import Autocomplete from "./Autocomplete";

configure({ adapter: new Adapter() });

describe("Autocomplete", () => {
  it("Render Autocomplete with default options", () => {
    const component = shallow(<Autocomplete name="autocomplete" />);
    const tree = shallowToJson(component);
    expect(tree).toMatchSnapshot();
  });
});
