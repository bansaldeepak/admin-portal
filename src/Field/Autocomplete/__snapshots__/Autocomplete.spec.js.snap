// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Render Autocomplete with default options 1`] = `
<WithStyles(ForwardRef(Autocomplete))
  filterOptions={[Function]}
  getOptionLabel={[Function]}
  getOptionSelected={[Function]}
  loading={false}
  name="autocomplete"
  onChange={[Function]}
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
  options={Array []}
  renderInput={[Function]}
/>
`;
